scripts/corelli/calibration/utils.py

```diff
@@ -75,6 +79,59 @@ def bank_numbers(bank_selection: str) -> List[str]:
     return [str(n) for n in sorted(banks)]
 
 
+def preprocess_banks(input_workspace: str, output_workspace: str) -> Workspace2D:
+    r"""
+    Clone the input workspace/run and preprocess the histograms for each tube such
+    that the peak finding algorithm can have a better chance of finding the correct
+    wire position.
+    Gaussian filters are used to approximate the background of each bank as well as
+    to assist the selection of regions where wire shadows reside.
+
+    Note: the preprocess step needs to be performed after load_banks, which provided
+    the bank selection.
+    :param input_workspace: input workspace name
+    :param output_workspace: output workspace with pre-processed histograms
+    """
+    CloneWorkspace(InputWorkspace=str(input_workspace), OutputWorkspace=output_workspace)
+
+    # inline function for 1D singal cleaning
+    # NOTE: an inline function here is by choice as it should only be used for data
+    #       pre-processing only
+    def clean_signals(
+        signal1D: np.ndarray,
+        pixels_per_tube: int = PIXELS_PER_TUBE,
+        peak_interval_estimate: int = WIRE_GAP_IN_PIXELS,
+    ) -> np.ndarray:
+        r"""
+        Replace the regions between peaks/dips with flat background to prevent peak finding
+        algorithm got stuck in a local minimum where no shadow of wires reside.
+
+        :param signal1D: 1D histogram from a single tube
+        :param pixels_per_tube: number of pixels per tube, should always be 256
+        :param peak_interval_estimate: a rough estimate of the distance between peaks in pixels
+        """
+        _sig_gaussian = gaussian_filter(signal1D, int(peak_interval_estimate / 2))
+        _sig_tmp = _sig_gaussian - signal1D
+        _sig_tmp[_sig_tmp < 0] = 1
+        _idx = np.where(_sig_tmp == 1)[0]
+        if len(_idx) > 0:
+            # This is mostly bypassing the non-realistic testing workspace
+            # which has zero values in most tubes
+            _sig_tmp[:_idx[0]] = 1
+            _sig_tmp[_idx[-1]:] = 1
+        _base = np.average(gaussian_filter(signal1D, int(pixels_per_tube / 2)))
+        return _base - _sig_tmp
+
+    _ws = mtd[output_workspace]
+    for i in range(0, _ws.getNumberHistograms(), PIXELS_PER_TUBE):
+        _data = np.array([_ws.readY(me) for me in range(i, i + PIXELS_PER_TUBE)])
+        _data = clean_signals(_data)
+        for j in range(PIXELS_PER_TUBE):
+            _ws.setY(i + j, _data[j])  # This apprently is the correct way to update Y
+
+    return _ws
```

Comment:
I'm just curious here. Why did you preprend every single variable with the private modifier `_` ?

Reply:
- Since I don't have a preference in Python code formatting, I am using `yapf` to automate the process whenever flake8 complains.  btw, `yapf` is also the recommend Python formatting tool for Mantid according to this doc https://github.com/mantidproject/mantid/blob/b70d0b9224eaf9ab542c80556fbe08b91cc0d118/dev-docs/source/Standards/PythonStandards.rst.  
- I double-checked my local config directory and there is no `.flake8` file, so I am probably using the default settings.
- As for the prefix, I was told by someone to follow this convention for local variable when I started at ORNL, but I can't remember who is the source :man_shrugging: